File: _includes/sort-benchmark/algorithms/self_indexed.rs
fn self_indexed_sort(a: &mut [usize]) {
    if a.is_empty() {
        return;
    }

    let min = *a.iter().min().unwrap();
    let max = *a.iter().max().unwrap();
    let span = max - min + 1;

    // Phase 1: initialize sorting space
    let mut ss = vec![0usize; span];

    // Phase 2: self-indexed arrangement (key maps to offset in ss)
    for &x in a.iter() {
        ss[x - min] += 1;
    }

    // Phase 3: order-preserved compression back into the original array
    let mut idx = 0;

    for (offset, &cnt) in ss.iter().enumerate() {
        let value = min + offset;
        for _ in 0..cnt {
            a[idx] = value;
            idx += 1;
        }
    }
}

File: _includes/sort-benchmark/algorithms/self_indexed.swift
func self_indexed_sort(_ a: inout [Int]) {
    a.withUnsafeMutableBufferPointer { self_indexed_sort($0) }
}

func self_indexed_sort(_ a: UnsafeMutableBufferPointer<Int>) {
    if a.isEmpty {
        return
    }

    var min = a[0]
    var max = a[0]
    for i in 1..<a.count {
        if a[i] < min { min = a[i] }
        if a[i] > max { max = a[i] }
    }
    let span = max - min + 1

    // Phase 1: initialize sorting space
    var ss = [Int](repeating: 0, count: span)

    // Phase 2: self-indexed arrangement (key maps to offset in ss)
    for i in 0..<a.count {
        ss[a[i] - min] += 1
    }

    // Phase 3: order-preserved compression back into the original array
    var idx = 0

    for offset in 0..<ss.count {
        let cnt = ss[offset]
        let value = min + offset
        for _ in 0..<cnt {
            a[idx] = value
            idx += 1
        }
    }
}
